### PR TITLE
Modgobi 150

### PR DESCRIPTION
**Pull request by SinghAdes**
## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by SinghAdes** [comment]
image

File: src/main/java/org/folio/gobi/Mapper.java [hunk 1]
```diff
@@ -720,18 +720,18 @@ private void mapLicense(List<CompletableFuture<?>> futures, License license, Doc
 
   private void mapFundDistribution(List<CompletableFuture<?>> futures, FundDistribution fundDistribution, Document doc,
                                    PostGobiOrdersHelper postGobiOrdersHelper) {
+    Optional.ofNullable(mappings.get(Mapping.Field.FUND_CODE))
```
Comment (grabsefx) [inline src/main/java/org/folio/gobi/Mapper.java:723]: It makes sense to map fund code only when fundId can be mapped. So this piece of code should be moved back

**Review by grabsefx — COMMENTED** [review]
It think the root cause of the issue lives here 
https://github.com/folio-org/mod-gobi/pull/161/files#diff-72855fef792ab800ef7faf83257793e7e030aebd35c87c0bb956a4e24de6964dL732

I suppose this mapping shouldnt contain any calls to third party systems, all required  data should be already prepared. Please look how is lookup functionality implemented in other places of gobi

File: src/main/java/org/folio/gobi/Mapper.java (hunk 1 above)
Comment (SinghAdes) [reply src/main/java/org/folio/gobi/Mapper.java]: can we make **" map fund id only when fundCode can be mapped"**?? as fundCode is also a mendatory field. Kindly suggest.

**Comment by SinghAdes** [comment]
image

File: src/main/java/org/folio/gobi/Mapper.java [hunk 2]
```diff
@@ -729,8 +728,8 @@ private void mapFundDistribution(List<CompletableFuture<?>> futures, FundDistrib
 
             Optional.ofNullable(mappings.get(Mapping.Field.FUND_CODE))
               .ifPresent(fundCodeField -> futures.add(fundCodeField.resolve(doc)
-                .thenCompose(fundCodeObject -> fundCodeResolver(fundDistribution, doc, postGobiOrdersHelper, (String) fundCodeObject))
-                .exceptionally(Mapper::logException)));
+                .thenApply(fundCodeObject -> fundCodeResolver(fundDistribution, doc, postGobiOrdersHelper, (String) fundCodeObject)
```
Comment (AndreiMPV) [inline src/main/java/org/folio/gobi/Mapper.java:730]: Should be thenCompose, because fundCodeResolver(...) returns ComplitableFututre

**Review by AndreiMPV — CHANGES_REQUESTED** [review]

File: src/main/java/org/folio/gobi/DataSourceResolver.java [hunk 3]
```diff
@@ -86,7 +86,7 @@ public static Builder builder() {
     if (translation != null) {
       return translation.apply(o == null ? null : o.toString());
     } else {
-      return CompletableFuture.completedFuture(o);
+      return CompletableFuture.supplyAsync(()->o);
```
Comment (AndreiMPV) [inline src/main/java/org/folio/gobi/DataSourceResolver.java:89]: Why you change that?

File: src/main/java/org/folio/gobi/DataSourceResolver.java [hunk 4]
```diff
@@ -45,9 +45,9 @@ public static Builder builder() {
     CompletableFuture<Object> future = new CompletableFuture<>();
 
     if (from != null) {
-      applyXPath(doc)
-        .thenAccept(s -> applyTranslation(s)
-          .thenAccept(o -> applyDefault(o, doc)
+      CompletableFuture.completedStage(applyXPath(doc)
```
Comment (AndreiMPV) [inline src/main/java/org/folio/gobi/DataSourceResolver.java:48]: Also what is the reason to add CompletableFuture.completedStage() here?
applyXPath(doc) already returned CompletableFuture

**Pull request closed without merging**